[SPARK-51069][SQL] Add big-endian support to UnsafeRowUtils.validateStructuralIntegrityWithReasonImpl

jonathan-albrecht-ibm opened the pull request:
### What changes were proposed in this pull request?
In UnsafeRowUtils.validateStructuralIntegrityWithReasonImpl, when checking if the unused bits in the field are all zeros for fixed length fields, add functions that work on big endian platforms.

### Why are the changes needed?
The current checks only work on little endian platforms. This change detects the platform endianness once when the getPadding* functions are initialized and sets it to the proper implementation for the platform so there should be no significant change in performance.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
This was tested on amd64 (little-endian) and s390x (big-endian) with existing unit tests. On s390x, many existing tests fail in the SQL projects.

### Was this patch authored or co-authored using generative AI tooling?
No

jonathan-albrecht-ibm committed:
In validateStructuralIntegrityWithReasonImpl when checking if the unused bits
in the field are all zeros for fixed length fields, add functions that work
on big endian platforms.

The current checks only work on little endian platforms. This change detects
the platform endianness once when the getPadding* function is initialized and
sets it to the proper implementation for the platform so there should be no
significant change in performance.

Signed-off-by: Jonathan Albrecht <jonathan.albrecht@ibm.com>

jonathan-albrecht-ibm committed:
In validateStructuralIntegrityWithReasonImpl when checking if the unused bits
in the field are all zeros for fixed length fields, add functions that work
on big endian platforms.

The current checks only work on little endian platforms. This change detects
the platform endianness once when the getPadding* function is initialized and
sets it to the proper implementation for the platform so there should be no
significant change in performance.

Signed-off-by: Jonathan Albrecht <jonathan.albrecht@ibm.com>

jonathan-albrecht-ibm commented:
If possible, I'd like to ask for this PR to be backported as far back as branch-3.5 if it is accepted. I have tested this fix on 3.5.4 on big and little endian platforms.

MaxGekk commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/UnsafeRowUtils.scala:
```diff
@@ -74,23 +92,23 @@ object UnsafeRowUtils {
         field.dataType match {
           case BooleanType =>
-            if ((row.getLong(index) >> 1) != 0L) {
+            if (getPaddingBoolean(row, index) != 0L) {
```
I worry about creating functions in the loop `expectedSchema.fields.zipWithIndex.foreach {`. This might cause perf regression. How about to place `getPadding` and `getPaddingBoolean` in separate classes like `UnsafeRowBeUtils` and `UnsafeRowLeUtils`, and instantiate them out of the loop.

jonathan-albrecht-ibm replied:
Thanks for the review @MaxGekk! 

The `getPadding*` functions are actually created outside of the loop already. They are declared as `val` members and `UnsafeRowUtils` is a top level `object` which makes it a singleton so the `getPadding*` functions are initialized only once for the entire program; not every time they are called. They are initialized with one of the anonymous functions returned by the `if` statement. So when the `getPadding*` functions are called, only the anonymous function chosen at initialization is executed. Only a single instance of the chosen anonymous function is ever created.

It's a really nice concise short form for creating separate classes like `UnsafeRowBeUtils` and `UnsafeRowLeUtils` and choosing one of those to use ahead of time.

jonathan-albrecht-ibm replied:
Hi @MaxGekk, just wanted to follow up on this. I hope my explanation makes sense.

jonathan-albrecht-ibm replied:
Hi @MaxGekk, I just wanted to follow up with you on my comments above.

jonathan-albrecht-ibm replied:
Hi @MaxGekk, just wanted to ask if you have some time to look at my comments above

jonathan-albrecht-ibm replied:
Hi @MaxGekk, just wanted to follow up again to ask if you have some time to look at my comments above

jonathan-albrecht-ibm replied:
Hi @dongjoon-hyun, this PR has been stuck for a while. You helped me with some of my other PRs. I was wondering if you had some time to look at this one or point me to someone who can review this area of the code.

jonathan-albrecht-ibm replied:
Hi @dongjoon-hyun, @MaxGekk, just wondering if anyone has some time to look at my comments above

dongjoon-hyun reviewed (changes requested):
Hi, @jonathan-albrecht-ibm . I also agree with the previous concerns which blocks this PR. The current proposal looks like a code-level performance regression to me. Please avoid introducing a new function to the existing code path if you want to make a progress.

For example,
1. `nativeOrder().equals(BIG_ENDIAN)` could be some `lazy val` variable somewhere instead of invoking at every row.
2. You may add new code in the following style before the existing code.

```scala
- case BooleanType =>
+ case BooleanType if newFlag =>
```

jonathan-albrecht-ibm committed:
Convert top level padding functions to object methods

Signed-off-by: Jonathan Albrecht <jonathan.albrecht@ibm.com>

jonathan-albrecht-ibm committed:
Convert top level padding functions to object methods

Signed-off-by: Jonathan Albrecht <jonathan.albrecht@ibm.com>

jonathan-albrecht-ibm commented:
Thanks @dongjoon-hyun, I am refactoring my changes as per @MaxGekk's suggestion https://github.com/apache/spark/pull/49773#discussion_r1947241109
I should have the new changes ready soon

jonathan-albrecht-ibm commented:
Thanks for reopening this PR @MaxGekk. I have made the changes you originally requested if you have time to review

MaxGekk commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/UnsafeRowUtils.scala:
```diff
@@ -17,11 +17,38 @@
 
 object UnsafeRowUtils {
 
+  sealed trait PaddingProvider {
```
Could you make the added trait/objects private (or private[util] if you want them reachable from the suite). They're an internal dispatch detail of one private method.

jonathan-albrecht-ibm replied:
Thanks, done

MaxGekk commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/UnsafeRowUtils.scala:
```diff
@@ -17,11 +17,38 @@
+      row: UnsafeRow, index: Int): Long = row.getLong(index) & 0xFEFFFFFFFFFFFFFFL
+  }
+
+  private final val padder: PaddingProvider = if (nativeOrder().equals(BIG_ENDIAN)) {
```
Isn't **final** redundant on a val? Also cosmetic:

```suggestion
  private final val padder: PaddingProvider = if (nativeOrder() == BIG_ENDIAN)) {
```

jonathan-albrecht-ibm replied:
Thanks, done

MaxGekk reviewed (commented):
Spark CI does not run on s390x. A small unit test calling PaddingProviderLE/PaddingProviderBE directly with hand-built Longs (clean + corrupt padding for each of the 5 types) would be feasible on any LE dev machine.

jonathan-albrecht-ibm committed:
Add unit test for PaddingProvider

Signed-off-by: Jonathan Albrecht <jonathan.albrecht@ibm.com>

jonathan-albrecht-ibm committed:
Address review comments

- make PaddingProvider traits private[util]
- nativeOrder().equals(BIG_ENDIAN) => nativeOrder() == BIG_ENDIAN

Signed-off-by: Jonathan Albrecht <jonathan.albrecht@ibm.com>

jonathan-albrecht-ibm committed:
Add test cases for corrupted fields

Signed-off-by: Jonathan Albrecht <jonathan.albrecht@ibm.com>

MaxGekk commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/UnsafeRowUtilsSuite.scala:
```diff
@@ -157,4 +158,119 @@ class UnsafeRowUtilsSuite extends SparkFunSuite {
+    val byteBE = Array[Byte](0xFF.toByte, 0, 0, 0, 0, 0, 0, 0)
+    val shortBE = Array[Byte](0xFF.toByte, 0xFE.toByte, 0, 0, 0, 0, 0, 0)
+    val intBE = Array[Byte](0xFF.toByte, 0xFE.toByte, 0xFD.toByte, 0xFC.toByte, 0, 0, 0, 0)
+    val floatBE = Array[Byte](0xFF.toByte, 0x7F.toByte, 0x7D.toByte, 0xFC.toByte, 0, 0, 0, 0)
```
Probably, a typo in  floatBE / floatCorruptBE.

The comment describes the float value as 0xFF7FFDFC. The correct big-endian byte layout is [FF, 7F, FD, FC, ...], but the arrays have [FF, 7F, 7D, FC, ...] - byte index 2 is 0x7D instead of 0xFD.

jonathan-albrecht-ibm replied:
Fixed, thx

jonathan-albrecht-ibm committed:
Fix float*BE values to match comment

Signed-off-by: Jonathan Albrecht <jonathan.albrecht@ibm.com>

jonathan-albrecht-ibm committed:
Fix linter issues

Signed-off-by: Jonathan Albrecht <jonathan.albrecht@ibm.com>
